addons/web/static/src/views/fields/section_note_list/section_note_list_field.js

```diff
@@ -0,0 +1,387 @@
+    }
+
+    get id() {
+        return this.record.id;
+    }
+
+    get index() {
+        const prop = this.isSection ? "childSections" : "childItems";
+        return this.parent[prop].indexOf(this) + 1;
+    }
+
+    get isLastChildItem() {
+        return this.index === this.parent.childItems.length;
+    }
+
+    get isNote() {
+        return this.type === DISPLAY_TYPES.NOTE;
+    }
+
+    get isRoot() {
+        return this.type === DISPLAY_TYPES.ROOT;
+    }
+
+    get isSection() {
+        return isSection(this.type);
+    }
+
+    get isTopSection() {
+        return this.type === DISPLAY_TYPES.SECTION;
+    }
+
+    get records() {
+        if (this.isRoot) {
+            return [
+                ...this.childItems.flatMap((it) => it.records),
+                ...this.childSections.flatMap((it) => it.records),
+            ];
+        }
+        if (this.isSection) {
+            return [
+                this.record,
+                ...this.childItems.flatMap((it) => it.records),
+                ...this.childSections.flatMap((it) => it.records),
+            ];
+        }
+        return [this.record];
+    }
+
+    get sectionValue() {
+        if (!this.isSection) {
+            return 0;
+        }
+        if (this.isTopSection) {
+            return this.index * TOP_SECTION_VALUE_FACTOR;
+        } else {
+            return this.parentValue + this.index * SUB_SECTION_VALUE_FACTOR;
+        }
+    }
+
+    addChild(item) {
+        const prop = item.isSection ? "childSections" : "childItems";
+        this[prop].push(item);
+    }
+}
+
+class SectionListRenderer extends ListRenderer {
+    static props = [
+        ...super.props,
+        "isM2M",
+        "displayTypeField",
+        "noteContentField",
+        "sectionContentField",
+    ];
+    static recordRowTemplate = "web.SectionListRenderer.Row";
+
+    setup() {
+        super.setup();
+        this.controls = [
+            ...this.controls,
+            {
+                type: "create",
+                string: _t("Add section"),
+                context: this.makeSectionCreationContext(false),
+            },
+            {
+                type: "create",
+                string: _t("Add note"),
+                context: this.makeNoteCreationContext(),
+            },
+        ];
+        this.itemMap = {};
+    }
+
+    /**
+     * @param {HierarchyItem} item
+     * @param {boolean} isSubSection
+     */
+    async addRowInSection(item, isSubSection) {
+        await this.props.list.leaveEditMode();
+        await this.props.list.model.mutex.exec(async () => {
+            const moveAfterRecord = (
+                isSubSection
+                    ? item.records
+                    : [item.record, ...item.childItems.map((it) => it.record)]
+            ).at(-1);
+            const context = isSubSection ? this.makeSectionCreationContext(true) : {};
+            const newRecord = await this.props.list._createNewRecordDatapoint({
```

Comment:
Private functions shouldn't be called from outside the model. It would make sense to have a `addNewRecordAtIndex`, or to support more `position` values than "top" or "bottom", like an index, in StaticList.